core/src/main/scala/quasar/destination/s3/impl/DefaultUpload.scala

```diff
@@ -44,17 +46,17 @@ import software.amazon.awssdk.services.s3.model.{
 final case class DefaultUpload[F[_]: Concurrent: ContextShift](client: S3AsyncClient, partSize: Int)
     extends Upload[F] {
 
-  def upload(bytes: Stream[F, Byte], bucket: Bucket, key: ObjectKey): F[Unit] =
-    Concurrent[F].bracketCase(startUpload(client, bucket, key))(createResponse =>
-      for {
-        parts <- uploadParts(client, bytes, createResponse.uploadId, partSize, bucket, key)
-        _ <- completeUpload(client, createResponse.uploadId, bucket, key, parts)
-      } yield ()) {
+  def upload(bytes: Stream[F, Byte], bucket: Bucket, key: ObjectKey): Stream[F, Unit] =
+    Stream.bracketCase(startUpload(client, bucket, key)) {
       case (createResponse, ExitCase.Canceled | ExitCase.Error(_)) =>
         abortUpload(client, createResponse.uploadId, bucket, key).void
       case (_, ExitCase.Completed) =>
         Concurrent[F].unit
-    }
+    } flatMap (createResponse =>
+      for {
+        parts <- uploadParts(client, bytes, createResponse.uploadId, partSize, bucket, key)
+        _ <- Stream.eval(completeUpload(client, createResponse.uploadId, bucket, key, parts))
```

Comment:
Do you mean to call `completeUpload` once? How many elements will `uploadParts` contain?